Rewritten Route Drawer

RoiEXLab opened the pull request:
This is finally a Good version of my attempts to rewrite the route drawer...
Before:
![Before](https://cloud.githubusercontent.com/assets/8350879/15557724/44f40582-22d5-11e6-896d-fb8ca8f2da36.png)
Now:
![After 1](https://cloud.githubusercontent.com/assets/8350879/15557731/51f29f50-22d5-11e6-8bb1-c1b3a66f6321.png)
![After 2](https://cloud.githubusercontent.com/assets/8350879/15557732/5483294c-22d5-11e6-9f0e-dbd796e2c763.png)
This rewrite requires the apache commons math library available on maven central, licensed under the Apache License, Version 2.0.
In comparison to my previous attempts this is well performing even on long distances like the example in 270BC

What do you think?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/751)
<!-- Reviewable:end -->

DanVanAtta commented:
Screenshots look good.
- Need some tests. Decompose as necessary so you can test things.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,193 @@
 
   /**
-   * Draw m_route to the screen, do nothing if null.
+   * Draws the route to the screen, does nothing if null.
```
If what exactly is null? 
Can we avoid calling this method altogether with null args?

RoiEXLab replied:
I think the route is meant... Kept this comment from the original route drawer...

DanVanAtta replied:
Since the code was touched, and you've effectively re-written it, please update it so it makes sense. It is a very common bug/defect/mistake to not update commentary when the code is touched.

But ultimately, please focus on code updates to address the review commentary. Once this is merged, chances are good all commentary in this PR will be 100% useless. The value is in fixing the code so that someone reviewing the code independently would not have all these same questions / concerns.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,193 @@
+      return;
+    }
+
+    final Point[] points = getRoutePoints(routeDescription, mapData);
```
Avoid declaring all variables up front. This is not very cohesive. Cohesion keeps related entities close together, declaration + usage. As is, I have to search through the code base to see where these are used and to determine theri scope.
Easiest way to do this, cut the line out, see where you get the first compiler error, and insert the line above the line generating the error. Please do this for all the variables.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,193 @@
+    graphics.setPaint(Color.red);
+    graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+
+    if (Arrays.asList(points).contains(null)) {// If the Array is null at some point
```
This looks like a bug, when would this happen? What is the UI behavior when we run into this?

RoiEXLab replied:
I think this is a legacy code snippet when I ran into errors with the points array...
Can be removed I guess...

RoiEXLab commented:
> - Need some tests. Decompose as necessary so you can test things.
So I'm going to add another testClass in which directory?

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,193 @@
+      drawMoveLength(graphics, routeDescription, points, xOffset, yOffset, view, mapData.scrollWrapX(),
+          mapData.scrollWrapY(), numTerritories, movementLeftForCurrentUnits);
+    }
+    if (numTerritories <= 1 || points.length <= 2) {
```
lots of complexity, what do these boolean conditions represent logically? 
Note: please do not answer that question in a PR comment, but have the code answer it for itself, so it is answered once and for all.

DanVanAtta commented:
> So I'm going to add another testClass in which directory?

Tests go into the 'test' folder under a package whose name matches the package of the class being tested.

DanVanAtta commented:
Note, the tests will create immense pressure for the design of this code to be clean. That is really the reason why you are writing the test, and where most of the benefit is gained. It is a *side effect* that you will so happen to have a regression test in place once you are done. The clean design will ensure that the cost for adding another feature, or manipulating this code, will stay relatively fixed. Furthermore, code with poor design and low test coverage or no test coverage at all has high correlation with defects. 

"Simplicity is prerequisite for reliability." 
- Dijkstra

DanVanAtta commented:
Another point to help motivate you, a test should be the *simplest* possible client of the code it is testing. If that simplest possible client is really complicated, then it is a sure sign that trouble is ahead.

RoiEXLab commented:
Ok I will do so later today... Thanks for your input... It's 0:25 am I'm probably going to bed now^^
Feel free to merge my other PR ^^

RoiEXLab commented:
So I started adding some test for the logic behind the route drawer...
As I started creating those test I found some possible causes for a indexoutofboundsexception or similar... (e.g. `array[0]` was set without checking if `array.length > 0`)
Also I enhanced the logic behind the getCoords method. It should be more accurate now.
The problem with this enhancement is, that I'm missing factor unknown in order to check if this calculation is right... (I know it is, but if it gets changed in the future)
I would appreciate it if someone could help me there...
Also if you have Ideas for other test please let me know...

frigoref commented:
Don't really understand what you are getting at. Could you explain in more detail how I can assist you?
Also, I am assuming you have taken back this topic and I can focus on other things?

RoiEXLab commented:
@frigoref you are right... I found a way to make the route look good and perform well... 
I'm mostly done with this PR, the only thing missing are some tests...
If you have some ideas for some tests feel free to let me know, otherwise you can do whatever you want...

RoiEXLab commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 31 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/751#-KIdxvIqZvnAWo_-9W6h-r1-31:-KIxpOC0kkLxu6Gz4q12:-499146009) ([raw file](https://github.com/triplea-game/triplea/blob/7e0b94f600032fab1c60f85b16971337e006d252/src/games/strategy/triplea/ui/MapRouteDrawer.java#L31)):*
<details><summary><i>Previously, DanVanAtta (Dan Van Atta) wrote…</i></summary>
> Since the code was touched, and you've effectively re-written it, please update it so it makes sense. It is a very common bug/defect/mistake to not update commentary when the code is touched.
> 
> 
> 
> But ultimately, please focus on code updates to address the review commentary. Once this is merged, chances are good all commentary in this PR will be 100% useless. The value is in fixing the code so that someone reviewing the code independently would not have all these same questions / concerns.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/751)*
<!-- Sent from Reviewable.io -->

RoiEXLab commented:
I added some test...
I have still no Idea, how I'm going to calculate the unknown factor - Declared as `@Ignore` right now...
If someone has an idea for this, please tell me otherwhise I will probably add this in a future PR if  there are no objections...

RoiEXLab commented:
Ok I changed the calculation a little bit... now a delta value of 1 is used... not the best solution but it works... added a todo comment to change this...

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+    graphics.setStroke(new BasicStroke(3.5f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND));
+    graphics.setPaint(Color.red);
+    graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+    if (routeDescription == null) {
```
does not seem like graphics should be changed unless the method will actually do something. The check for null and return statements should probably be before we do anything else.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+    final double scale = view.getScale();
+    final int jointsize = 10;
+    if (tooFewTerritories || tooFewPoints) {
+      drawLineWithTranslate(graphics, new Line2D.Float(routeDescription.getStart(), routeDescription.getEnd()), xOffset,
```
Seems like the level of abstraction is mixed in this method. `tooFewTerritories` is a pretty high level of abstraction, and below we have some pretty detailed drawing routines.  Can this method be broken up into helper methods? (the goal being to hide some of the details and keep this function to be relatively high level)

DanVanAtta replied:
Looking at this for a second longer, I recommend collapsing the lines 54-56 together into the same helper method with a single parameter `routeDescription`, similar for the else part of thise `if/else`, but with slightly different params

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+      drawMoveLength(graphics, routeDescription, points, xOffset, yOffset, view, mapData.scrollWrapX(),
+          mapData.scrollWrapY(), numTerritories, movementLeftForCurrentUnits);
+    }
+    drawJoints(graphics, points, xOffset, yOffset, jointsize, scale);
```
Could use a comment here on what is meant by a 'joint'

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+    if (cursorImage != null) {
+      graphics.drawImage(cursorImage,
+          (int) (((routeDescription.getEnd().x - xOffset) - (cursorImage.getWidth(null) / 2)) * scale),
+          (int) (((routeDescription.getEnd().y - yOffset) - (cursorImage.getHeight(null) / 2)) * scale), null);
```
Similar to above comments, pushing this logic into a helper function could be nice, likely with two params `cursorImage` and `routeDescription`

RoiEXLab replied:
Outsourced that
Somehow not updating here...

DanVanAtta replied:
Check that you pushed to the correct branch. Sometimes good to verify the SHA of the commit you have local, against the commits in the branch:  https://github.com/triplea-game/triplea/pull/751/commits

RoiEXLab replied:
Oh I know what's going on... Git thinks you commented on the new method since its the exact same code...

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
-      final Point2D p2 = new Point2D.Double(line.getP2().getX() + translateX, line.getP2().getY() + translateY);
-      graphics.draw(new Line2D.Double(p1, p2));
+    for (int i = 1; i < points.length; i++) {
+      index[i] = index[i - 1] + Math.sqrt(points[i - 1].distance(points[i]));
```
could use a comment on what this formula represents. Specifically would want the name of the method or algorithm being used.
Also, any links to further readings?

RoiEXLab replied:
http://stackoverflow.com/a/37370620/5769952

DanVanAtta replied:
Please add that to the code if you would. 
Is there really no name for this technique/formula? (`index[i - 1] + Math.sqrt(points[i - 1].distance(points[i]));`)

frigoref replied:
I was just reading this code line and was wondering if the Math.sqrt is needed as I thought we would just take the distance and maybe it is a left over from the manually distance calculation we had before. Please recheck.
Otherwise, what kind of name were you, @DanVanAtta, referring to? The index values are providing some kind of weights based on index value differences to the point.X and point.Y values for the interpolation.

RoiEXLab replied:
It's not needed but I was told that this is working better...

DanVanAtta replied:
There is still a maintenance problem here. Someone looking at this line will have some difficult knowing why/what is being done: `index[i - 1] + Math.sqrt(points[i - 1].distance(points[i]));`


A comment or moving the calculation to a  one line helper function both could be ways to help that problem.  @RoiEXLab , can be taken up in a follow-on PR.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+
+  }
+
+  protected static double[] getIndex(Point[] points) {
```
The name of this method does not match the functionality. The name of the helper method can be a bit more detailed and longer to help describe what is happening.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+      points[i] = mapData.getCenter(territories.get(i));
     }
+    if (routeDescription.getStart() != null) {
+      points[0] = routeDescription.getStart();
```
Is there a way to guarantee that `territories` would include `routeDescription.getStart()` as the zero'th element always?

Worst case we could put a helper method on mapData: `mapData.getCenter(territories.get(i) , routeDescription );`, which then at least hide this type of detail from us. There might be a better / more direct fix though.

DanVanAtta replied:
Failing the above,  a comment to say why we are overwriting the first value would be useful.

RoiEXLab replied:
I could add a return statement if it is null...

DanVanAtta replied:
That could be preferable. The overwriting of the first element seems very hacky.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
-        controlDiffy = controlDiffy < 0 ? -maxControlLength : maxControlLength;
-        controlDiffx = controlDiffx < 0 ? (int) (-maxControlLength * ratio) : (int) (maxControlLength * ratio);
-      }
+  protected static Tuple<double[], double[]> pointsToDoubleArrays(Point[] points) {
```
I recommend avoiding using `Tuple`. Plus there is some redundancy in the code below. I think it can be fixed up with a helper method having this API:

`private static double[] getValues(Point[] point, java.function.Function<Point, Double> extractor)`

And you could call it like this:
```
double[] xValues = getValues(points, point -> point.getX())
double[] yValues = getValues(points, point -> point.getY())
```

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+  }
+
+  protected static double[] getCoords(PolynomialSplineFunction curve, double[] index) {
+    final double detailLevel = 1.0;//Tweak this value in order to achieve good rendering results
```
The comment here does not quite go far enough, I think it is not sufficient by itself.
Recommend to fix by extract the 1.0 to a named constant `private static final double MAX_DETAIL`, or something like that. Then a comment stating that it could be lowerd or raised to change rendering quality would be helpful/sufficient.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+    final int textXOffset = xDir > 0 ? 6 : (xDir == 0 ? 0 : -14);
+
+    final int yDir = cursorPos.y - points[numTerritories - 2].y;
+    final int textYOffset = yDir > 0 ? 18 : -24;
```
Please extract the numbers to named constants, on this line would be `18`, and `-24`. The goal would be to answer the question of what these number represent. There are quite a few of them in this function as well.

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+        movementLeftForCurrentUnits == null || movementLeftForCurrentUnits.trim().length() == 0 ? ""
+            : "    /" + movementLeftForCurrentUnits;
+    final BufferedImage movementImage = new BufferedImage(72, 24, BufferedImage.TYPE_INT_ARGB);
+    setupTextImage(movementImage.createGraphics(), String.valueOf(numTerritories - 1), unitMovementLeft);
```
pass in `movementLeftForCurrentUnits` to this method, and then you can hide the details of how to construct the movement left string in your helper function

RoiEXLab replied:
You mean unitMovementLeft right?

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+    graphics.drawImage(movementImage,
+        (int) ((cursorPos.x + textXOffset - xOffset - translateX) * scale),
+        (int) ((cursorPos.y + textYOffset - yOffset + translateY) * scale), null);
+    graphics.drawImage(movementImage,
```
Is the double drawing of the image intentional?Line 162 and 165 look to do the same thing.

RoiEXLab replied:
It is needed somehow don't ask me why

DanVanAtta replied:
something must be wrong if it is actually needed. Please have another look at this

RoiEXLab replied:
Ok it's not need at all, because scrollWrapX and Y are always false...
I had issues with this before, but whatever...
Also If you didn't notice, this method was actually called 3 times

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -6,295 +6,192 @@
+  private static void drawCurvedPath(Graphics2D graphics, Point[] points, MapPanel view) {
+    final double[] index = getIndex(points);
+    final Tuple<double[], double[]> pointArrays = pointsToDoubleArrays(points);
+    final PolynomialSplineFunction xcurve = splineInterpolator.interpolate(index, pointArrays.getFirst());
```
Cohesion could be better here. IE: similar things grouped closely together.

IE:
```
xCurve = ...
xCoords = ...

yCurve = ...
yCoords = ...
```

In fact, it looks like the y coordinate stuff is not used/needed until line 183. So the y stuff could be moved down a few lines.

DanVanAtta commented:
This is getting quite close   : )

Still needed:
- [ ] left some code review comments
- [ ] reviewer/additional testing

DanVanAtta commented on test/games/strategy/triplea/ui/TestRoute.java:
```diff
@@ -0,0 +1,64 @@
+package games.strategy.triplea.ui;
```
Good start on test cases.

Looking at these, the coverage does not appear very deep. Second thing I notice is that a lot of helper / protected methods are being tested. That is one way to test things, but increasing scope of a helper method so you can test it is a warning sign. In specific it then does not help you when you need to test the clients of those helper methods, you'll wind up with redundant testing.

I'd like to see us avoid the test of the helper methods.

One way to do this:
- create a package that will contain all the drawing logic / classes
- for each protected helper method, make a class out of it.
- Then test each class individually, and use mocks to stub out the dependencies. Then for example, you could mock out the 'getIndex()' functionality, and then your test could control the values coming from that method. Basically anywhere you have a protected method, you probably should have a mock.
- Then once you have things decomposed, your top level algorithm will actually be high level, and it'll be resistant to change since the details will be hidden from it, so it would only be a helper class that you would actually have to change if you update the drawing algorithm.

RoiEXLab replied:
But how would I check if something is drawn right?

DanVanAtta replied:
Separation of concerns. Isolate the code that does the drawing, test that manually. Basically break things up into classes, not just methods. A class can be mocked and tested independently. Eventually then the parts do the actual UI work, that interacts with the Graphics object, will be just about as small as possible, and it is is okay to test that manually.

RoiEXLab replied:
Problem here, currently all of those methods are static and mockito does not support static mocks...powermockito Could help out there - or a complicated refactoring

DanVanAtta replied:
Part of the problem is dependency injection, being able to insert mock objects as appropriate. Static is bad for that, creates some tight coupling.

Here is one thing you can do:
```
private static MyLogicHelperClass helperClass = new DefaultInstance();
protected static void setHelperClass(MyLogicHelperClass mockValue) {
   helperClass = mockValue;
}
```
There are still some problems there, but that would let you break up the static logic across multiple classes. Do avoid using powermockito to mock statics, generally there are easier/better ways to achieve that goal. Push comes to shove, powermockito is an option, but should be last resort.

RoiEXLab replied:
I also could change the method to non static... This way I could mock the MapRouteDrawer and call the drawing methods on the mock...

DanVanAtta replied:
That seems like an option as well :+1:

DanVanAtta commented on src/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -11,37 +11,44 @@
+   * Too low values make the Path look edgy, too high values will cause lag and rendering errors
+   * because the distance between the drawing segments is shorter than 2 pixels 
+   */
+  public static final double DETAIL_LEVEL = 1.0D;
```
The value "1.0D" can simply be 1.0
Floats and longs are the only types you really need to annotation with "f" or "L"

RoiEXLab commented:
Any suggestions for Test?

RoiEXLab commented:
I added another test, verfifying that the route is gotten etc.
Enough?

DanVanAtta commented:
@RoiEXLab  - more practice and testing gets to be a bit easier. The question is, if the code were not working, would the tests find out? Line coverage helps answer the question, and an interesting way to really check the thoroughness of the test is to insert random bugs into the code and verify that a test hits it.
